[libc++][hardening] Constrain construction for `__{(static_)bounded,wrap}_iter`

frederick-vs-ja opened the pull request:
This PR restricts construction to cases where reference types of source/destination iterators are (`T&`, `T&`) or (`T&`, `const T&`) ( where `T` can be const).

Fixes #50058.

frederick-vs-ja committed:
[libc++][hardening] Constrain construction and use `static_assert`

... for `__{bounded.wrap}_iter`.

This PR restricts construction to cases where reference types of
source/destination iterators are (`T&`, `T&`) or (`T&`, `const T&`) (
where T can be const).

We can't `static_assert` `__libcpp_is_contiguous_iterator` for
`__wrap_iter` currently because `__wrap_iter` is also used for wrapping
user-defined fancy pointers.

frederick-vs-ja commented:
Timed out in `std/utilities/charconv/charconv.msvc/test.pass.cpp`, which seems unrelated.

https://github.com/llvm/llvm-project/actions/runs/11718999636/job/32643141533

philnik777 commented:
Please make these separate PRs. The changes are unrelated AFAICT.

frederick-vs-ja committed:
Merge branch 'llvm:main' into restrict-bounded-wrap-iter

frederick-vs-ja committed:
Revert `static_assert`

frederick-vs-ja commented:
> Please make these separate PRs. The changes are unrelated AFAICT.

Separated to #115304.

ldionne commented on libcxx/include/__iterator/bounded_iter.h:
```diff
@@ -67,7 +72,13 @@ struct __bounded_iter {
-  template <class _OtherIterator, __enable_if_t< is_convertible<_OtherIterator, _Iterator>::value, int> = 0>
+  template < class _OtherIterator,
+             __enable_if_t<
+                 _And<is_constructible<_Iterator, const _OtherIterator&>,
+                      is_convertible<const _OtherIterator&, _Iterator>,
```
Do you need to check both `is_convertible` and `is_constructible`? Doesn't convertibility subsume constructibility?

frederick-vs-ja replied:
I'll change back to use `is_convertible` only...

However, [allocator.requirements.general] (and perhaps many other places in the library specification) sometimes only says "(implicit) conversion"/"(implicitly) convertible", while C++ types can be so tricky that explicit and implicit conversions can have different semantics, or the explicit conversion can be ill-formed while the implicit one is not. Maybe an LWG issue, requiring conversions between fancy pointers to be sane, is desired.

frederick-vs-ja commented on libcxx/include/__iterator/bounded_iter.h:
```diff
@@ -67,7 +72,13 @@ struct __bounded_iter {
+             __enable_if_t<
+                 _And<is_constructible<_Iterator, const _OtherIterator&>,
+                      is_convertible<const _OtherIterator&, _Iterator>,
+                      _Or<is_same<reference, __iter_reference<_OtherIterator> >,
+                          is_same<reference, __make_const_lvalue_ref<__iter_reference<_OtherIterator> > > > >::value,
```
Yes. [[iterator.concept.contiguous]](https://eel.is/c++draft/iterator.concept.contiguous) and [[allocator.requirements.general]/92](https://eel.is/c++draft/allocator.requirements.general#92) (for possibly fancy pointers, note that `std::addressof` rejects rvalues) require this.

frederick-vs-ja committed:
Merge branch 'main' into restrict-bounded-wrap-iter

frederick-vs-ja committed:
Address review comments

- Drop uses of `is_constructible`.
- Generalize to `__static_bounded_iter`.
- Improve conditional compilation in test.

frederick-vs-ja committed:
Fix old friend declaration

frederick-vs-ja commented:
> Can you also make this change for `__static_bounded_iter` which was added just now?

Changed. Also added a previously missed change to one friend declaration.
